### PR TITLE
Add client and server support for SMTP DSN extension

### LINKED ISSUE #32: DSN support
https://tools.ietf.org/html/rfc3461

### PR DESCRIPTION
**Pull request by foxcpp**
Breaking change since it changes Client.Rcpt and Session.Rcpt signature.

Closes #32.

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -400,12 +411,48 @@ func (c *Client) Mail(from string, opts *MailOptions) error {
...
+// to the command. Handling of unsupported options depends on the extension.
+//
 // If server returns an error, it will be of type *SMTPError.
-func (c *Client) Rcpt(to string) error {
+func (c *Client) Rcpt(to string, opts *RcptOptions) error {
 	if err := validateLine(to); err != nil {
 		return err
 	}
-	if _, _, err := c.cmd(25, "RCPT TO:<%s>", to); err != nil {
+	cmdStr := "RCPT TO:<%s>"
```
Comment (emersion) [inline client.go:422]: Is it safe to just append content to `cmdStr`? The new content could have user-supplied `%v` sequences, which `fmt` will choke on.

We should probably set `cmdStr = "RCPT TO:<%s>%s"` and have a separate `extStr` string.

File: conn.go [hunk 2]
```diff
@@ -478,20 +508,68 @@ func (c *Conn) handleRcpt(arg string) {
...
+			c.WriteResponse(501, EnhancedCode{5, 5, 1}, "Was expecting TO arg syntax of TO:<address>")
+			return
+		}
+	}
+	recipient := toArgs[0]
+	if recipient == "" {
+		c.WriteResponse(501, EnhancedCode{5, 5, 2}, "Was expecting MAIL arg syntax of FROM:<address>")
+		return
+	}
+	recipient = strings.Trim(recipient, "<>")
```
Comment (emersion) [inline conn.go:527]: Can we keep the TODO about this trim being not quite what we want?

File: conn.go [hunk 3]
```diff
@@ -478,20 +508,68 @@ func (c *Conn) handleRcpt(arg string) {
...
+
+	if len(toArgs) > 1 {
+		args, err := parseArgs(toArgs[1:])
+		if err != nil {
+			c.WriteResponse(501, EnhancedCode{5, 5, 4}, "Unable to parse TO ESMTP parameters")
+			return
+		}
+
+		for key, value := range args {
+			switch key {
```
Comment (emersion) [inline conn.go:544]: Do we want to error out on unrecognized parameter?

File: conn.go [hunk 4]
```diff
@@ -478,20 +508,68 @@ func (c *Conn) handleRcpt(arg string) {
...
+			case "ORCPT":
```
Comment (emersion) [inline conn.go:545]: Can we add a `// This space is intentionally left blank` comment on the next line? IMHO it's too easy to misread this as a fall-through.

**Comment by emersion** [comment]
Do you have plans to add basic test cases for these new features?

**Comment by foxcpp** [comment]
> Do you have plans to add basic test cases for these new features?

Yeah, they should be added.

File: client.go (hunk 1 above)
Comment (foxcpp) [reply client.go:422]: It applies to MAIL FROM too which uses similar code.

File: backend.go [hunk 5]
```diff
@@ -71,7 +122,7 @@ type Session interface {
 	// Set return path for currently processed message.
 	Mail(from string, opts MailOptions) error
 	// Add recipient for currently processed message.
-	Rcpt(to string) error
+	Rcpt(to string, opts RcptOptions) error
```
Comment (kayrus) [inline backend.go:125]: If #145 is accepted, `opts` should be switched to a pointer as well.

File: conn.go (hunk 2 above)
Comment (kayrus) [reply conn.go:527]: I also don't like the existing TODO, and proposed to use `mail.ParseAddress`. How is it implemented in postfix/exim mail servers? Shall we introduce a similar function?

**Comment by ikedas** [comment]
Hi,

I just wrote a DSN extension also: https://github.com/emersion/go-smtp/compare/v0.18.0...ikedas:go-smtp:dsn_extension

I think my code is better in the following respects:

* It includes the fix for PR #232 (conversely, this should be merged first).
* Takes into account the case where SMTPUTF8 extension is enabled (update by RFC6533).

Should I submit a separate PR?

**Comment by emersion** [comment]
Please feel free to submit a new PR, yeah.

Do note that we can still do breaking changes to our API if we feel that they are worth it.

**Comment by emersion** [comment]
Superseded by #233

**Pull request closed without merging**